### PR TITLE
*: make gc not write db directly

### PR DESCRIPTION
**Pull request by BusyJay**
### What is changed and how it works?

Issue Number: Ref #12842 

What's Changed:

```commit-message
We rely on non-concurrent memtable write for dynamic regions to
achieve best performance. This PR makes sure writes of compaction
filter be redirected to apply thread when dynamic regions is enabled.

The solution may miss data if TiKV crashes before writes are flushed
to disk. Note even for v1, it's also possible to leave garbage if writes
to rocksdb fail. We need to scan default CFs and check for orphan
versions.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/server/gc_worker/rawkv_compaction_filter.rs [hunk 1]
```diff
@@ -135,8 +136,6 @@ impl Drop for RawCompactionFilter {
     fn drop(&mut self) {
         self.raw_gc_mvcc_deletions();
 
-        self.engine.sync_wal().unwrap();
```
Comment (tonyxuqqi) [inline src/server/gc_worker/rawkv_compaction_filter.rs:138]: sync_wal is not needed in v2. But what about v1? I suppose src/server/gc_worker/rawkv_compaction_filter.rs is for v1 as well.

File: src/server/gc_worker/gc_worker.rs [hunk 2]
```diff
@@ -871,6 +875,46 @@ impl<E: Engine> GcRunner<E> {
...
+                    break;
+                }
+                modifies.push(Modify::Delete(CF_DEFAULT, keys.next().unwrap()));
+            }
+            if !modifies.is_empty() {
+                region_modifies.insert(region.id, modifies);
+                modifies = vec![];
+            }
+        }
+        if let Err(e) = self.engine.modify_on_kv_engine(region_modifies) {
```
Comment (tonyxuqqi) [inline src/server/gc_worker/gc_worker.rs:914]: How to ensure when the apply runs the unsafe delete, it's still the same tablet. I think we need to add the tablet suffix in the region_modifies

**Review by tonyxuqqi — APPROVED** [review]

File: src/server/gc_worker/rawkv_compaction_filter.rs (hunk 1 above)
Comment (BusyJay) [reply src/server/gc_worker/rawkv_compaction_filter.rs:138]: rawkv doesn't write to kv db directly, it was added by accident.

File: src/server/gc_worker/gc_worker.rs (hunk 2 above)
Comment (BusyJay) [reply src/server/gc_worker/gc_worker.rs:914]: We don't. The deleted keys are the keys that's stale. Once they are stale, they will always be. So always delete them in the latest tablet is correct.

**Review by tabokie — APPROVED** [review]

**Pull request merged**